### DISCUSSION TITLE
Load a web worker with different public path

### DISCUSSION THREAD

**Discussion by gretzkiy** [webpack/webpack#16843] [Q&A] [answered]
Hello! I'm currently having difficulty creating a web-worker due to the same-origin restrictions. The assets of my web app are hosted on a CDN, which makes it impossible for me to load the web worker from the default asset public path. I am attempting to use the new syntax for creating a web worker with `new Worker(new URL('./worker.js', import.meta.url))` in Webpack 5, but it seems that when I use module assets rules for the worker, Webpack creates a second separate asset for the worker and doesn't change the path in the `new Worker(...)` statement in source code. Here are my main points:

1. I have a web worker with modules imports in it and it may have been written in TypeScript, although for now I'm just using JavaScript for simplicity.
2. I want the worker to be loaded from the public path that matches my web app's domain, rather than the CDN public path, to bypass same-origin restrictions.
3. I can create a proxy within my web app's server that routes the request to the real CDN, so I want to emit my workers to a specific output path inside my dist folder, such as workers/. It can look something like `https://mywebapp-domain.com/workers/worker1` -> `https://my-cdn.com/path-to-my-build-dist/workers/worker1`.

Is it possible to achieve this with the new Worker(new URL('./worker.js', import.meta.url)) syntax? I have created a simple reproduction case in this Github repository: https://github.com/gretzkiy/webpack-url-assets-test/tree/master/project. Thank you for your help!

**Comment by alexander-akait** [comment]
Are you looking for https://github.com/webpack/webpack/pull/16671?

**Reply by gretzkiy** [reply]
Yeah, looks like that, thanks! But it seems that I’m missing something with the configuration of asset modules, because my worker isn’t loaded at all even from CDN domain. Can you please give me a hint whether my build is configured correctly or not in order to load compiled worker? Do I use asset/resource rule in the right way to emit a loader with custom file name?

**Reply by alexander-akait** [reply]
Why do you use `type: 'asset/resource'`? If you want to use it, you need to setup other loader here too, like babel/typescript/etc, we have two solutions for reccomendations:
- Use `new Worker(...)` syntax (but I don't think it fits you)
- Multi compiler mode:
```js
module.exports = [
        {
          target: ['webworker', 'es2022'],
          // Other options
        },
	{
		name: 'app',
		target: ['web', 'es2022'],
		mode: 'development',
		entry: path.resolve(__dirname, 'src', 'index.js'),
		output: {
			path: path.resolve(__dirname, 'dist'),
		},
		devtool: 'source-map'
	}
];
```

And if your application dependes on a worker logic (i.e. used it in code), please use https://webpack.js.org/configuration/other-options/#dependencies

**Reply by gretzkiy** [reply]
I apologize for oversimplifying my previous example. In fact I have a library that can spawn multiple web workers, and I am trying to find a way to create a general webpack rule that will process all of the workers in my final application (which uses this library), without having to specify a separate build for each worker or know the entry path for each worker.

**Reply by alexander-akait** [reply]
> without having to specify a separate build for each worker

I don't think it is possible because your library cosume worker files... Anyway maybe you can provide a reproducible example so I take a look and try to find the best solution

**Reply by gretzkiy** [reply]
Sure, here it is. https://github.com/gretzkiy/webpack-url-assets-test

I've been able to achieve almost everything I wanted, but the final step is to set a custom output path for all workers, so that they can be placed in `dist/worker` instead of just `dist` (that's why I tried to use `asset/resource` with custom output path in my first example). 

Here I load the workers inside the library: https://github.com/gretzkiy/webpack-url-assets-test/blob/957b0a570b193752ce8fd51fbe4830d4b7a6490a/lib/src/index.ts#L6

The library is then used in the project: https://github.com/gretzkiy/webpack-url-assets-test/blob/957b0a570b193752ce8fd51fbe4830d4b7a6490a/project/src/index.ts#L3 

In the project's webpack configuration file, I have added the new `workerPublicPath` option: https://github.com/gretzkiy/webpack-url-assets-test/blob/957b0a570b193752ce8fd51fbe4830d4b7a6490a/project/webpack.config.js#L14

However, I'm having trouble figuring out how to place the processed worker files within a custom output path. I believe I can accomplish this using a webpack plugin, but I was wondering if you could suggest a better approach that uses native webpack tooling?

**Reply by alexander-akait** [reply]
Sorry for delay, I always answer on questions, but sometimes I have a lot of them so I have delay, do you need this:
```js
'use strict';

const path = require('path');

module.exports = [
	{
		name: 'app',
		target: ['web', 'es2022'],
		mode: 'development',
		entry: path.resolve(__dirname, 'src', 'index.ts'),
		output: {
			path: path.resolve(__dirname, 'dist'),
			publicPath: 'https://my-cdn.com/path-to-my-build-dist/',
			workerPublicPath: '/workers/',
		},
		module: {
			rules: [
				{
					test: /\.svg/,
					type: 'asset/resource',
					generator: {
					        // It can be absolute URL with `http`/`https`
						publicPath: '/dist/',
					},
				},

				{
					test: /\.ts$/,
					use: 'ts-loader',
				},
			]
		},

		resolve: {
			extensions: ['.ts', '.js'],
		},

		devtool: 'source-map'
	}
];
```

look here https://webpack.js.org/configuration/module/#rulegenerator, there are a lot of options, many of them were developed for worker purposes, special https://webpack.js.org/configuration/module/#rulegeneratorpublicpath and https://webpack.js.org/configuration/module/#rulegeneratoroutputpath

I lost a bit of context... Do you store assets on CDN, right? Do you have workers locally or on CDN too? Do you want to load assets only from appliction (or from workers too)? May you show me your file strcuture of your application (simplified) and there you can have a problem with loading. I looked at your repo, but I don't undestand you want to proxy local requests for assets or do you want to avoid it (or it is just for an example purpose). I think you don't need a plugin and can solve it using webpack tooling

Also feel free to write to me in telegram (link on my profile page)

**Comment by gretzkiy** [comment, marked as answer]
@alexander-akait assisted me in troubleshooting this issue, and the outcome is as follows:

1. Currently, there is no simple method to modify the chunk name exclusively for the worker.
2. A somewhat unreliable approach exists (outlined in the linked issue below), but it is not considered dependable.
3. We have submitted a feature request to introduce a new option that can be used to designate a chunk filename solely for workers: https://github.com/webpack/webpack/issues/17002.
